### ISSUE TITLE
More basic shapes for prototyping

### ISSUE THREAD

**Issue by LegenDAirie** [Zinggi/elm-2d-game#1]
Being able to use a few more built in shapes without having to use sprites or custom render functions would help make prototyping more productive because being limited to rectangles makes it difficult to distinguish between game entities.

##### Shapes that would help
Shape | Potential Use
--- | ---
Circle | Player
Triangle | Enemy

##### Nice to haves / potential ideas
Shape | Potential Use
--- | ---
Ring | Hitbox
Line | Path
Curved Line | Trajectory
Hexagon | Dangerous terrain
Octagon | Ally npc

I have not used veryCustom but it looks like the current way to implement other shapes is it use that. Is that correct?

In the Game.TwoD.Shapes section of the docs you specify that "Since we're dealing with 2d only, the only available shape is a square". Are the shapes I am listing imposible to implement?

Would these changes be in the right direction of this project?

**Comment by Zinggi** [comment]
That would be quite nice to have indeed.
Unfortunately I currently just don't have enough free time to develop these things, however, a pull request would be very welcome.

All of these things can actually be done using `customFragment`.
The idea is that you draw whatever shape you want and make the other pixels transparent.
Here is a tutorial on how to draw stuff with fragment shaders: https://www.youtube.com/watch?v=0ifChJ0nJfM
The tutorial draws quite a complex shape, but the principles for other shapes are the same.
I found a few shadertoy examples that demonstrates how to render these shapes: https://www.shadertoy.com/view/4dfXDn
https://www.shadertoy.com/view/XtjGzt
And here is a written tutorial on a few shapes:
https://thebookofshaders.com/07/



For some of the shapes you suggested (Triangle, Hexagon, Octagon), it might be easier to draw them with `veryCustom`, this way you could create the required shape with an appropriate list of vertices.


So if you have the time, please explore how to draw these shapes. If you get satisfactory results, you could either send a pull request, or create an extension library that provides these things.

**Comment by LegenDAirie** [comment]
Awesome! thanks for sharing all of that information.

I would love to contribute.

This is what I am thinking so far. The most straightforward way that pops out to me is to make a unitTriangle in the Shapes.elm file. From there make a new Renderable type called ColoredTriangle.
This should work well for the triangle since the only real different is the vertices in the unit(Shape) function. Unfortunately, this does not seem as straightforward for rings circles and lines. Would It be a bad idea to have some shapes implemented with vertices and then others using a fragment shader instead of just all in fragment shaders?

My goal is to make it ease to select from multiple basic shapes without bloating the project too much. Do you think a good approach would be to make a BasicShape type and replace ColoredRectangle with ColoredShape that take a Basic Shape as extra data? for example something along the lines of

```
type BasicShape =
    = Square
    | Triangle
    | Circle
    | Ring 
    | Line


type Renderable
    = ColoredShape { shape : BasicShape, transform : Mat4, color : Vec3 }
    | TexturedRectangle {stuff}
    | AnimatedSprite {stuff}
    | ParallaxScroll {stuff}
    | Custom (stuff)
```

Using this approach the implementation could be abstracted away and someone could just select the shape they wanted.

What are your thoughts?

**Comment by Zinggi** [comment]
> Would It be a bad idea to have some shapes implemented with vertices and then others using a fragment shader instead of just all in fragment shaders?

I think that's ok. I'd want the triangle to be implemented via a shape, but a perfectly round circle should be implemented via a fragment shader.

> Do you think a good approach would be to make a BasicShape type and replace ColoredRectangle with ColoredShape that take a Basic Shape as extra data?

Yep, that looks like a good idea to me.
But make sure you don't expose the constructors of the union type.
This gives us the freedom to extend the union at any time in a minor release.
If it was exposed, we'd have to bump the major version every time we add a new shape. 
So only expose `BasicShape`, not `BasicShape(..)` and create constructors like `circle = Circle` etc. and expose those.

> What are your thoughts?

Go ahead and send me a pull request when done :smile:

**Comment by Zinggi** [comment]
fixed in #2
